docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,879 @@
+spec:
+  cloudConfig: base64(downloader-cloud-config)
+  infrastructureProviderStatus:
+    apiVersion: aws.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureStatus
+    vpc:
+      id: vpc-1234
+      subnets:
+      - id: subnet-acbd1234
+        name: workers
+        zone: eu-west-1
+      securityGroups:
+      - id: sg-xyz12345
+        name: workers
+    iam:
+      nodesRoleARN: <some-arn>
+      instanceProfileName: foo
+    ec2:
+      keyName: bar
+  providerConfig:
+    apiVersion: aws.cloud.gardener.cloud/v1alpha1
+    kind: WorkerPoolConfig
+    machineImage:
+      name: CoreOS
+      ami: ami-d0dcef3b
+    machineType: m4.large
+    volumeType: gp2
+    volumeSize: 20Gi
+    zones:
+    - eu-west-1a
+  region: eu-west-1
+  secretRef:
+    name: my-aws-credentials
+  minimum: 2
+  maximum: 2
+status:
+  observedGeneration: ...
+  state: ...
+  lastError: ..
+  lastOperation: ...
+```
+
+#### Shoot state
+
+In order to enable moving the control plane of a Shoot between Seed clusters (e.g., if a Seed cluster is not available anymore or entirely broken) Gardener must store some non-reconstructable state, potentially also the state written by the controllers.
+Gardener watches these extension CRDs and copies the `.status.state` in a `ShootState` resource into the Garden cluster.
+Any observed status change of the respective CRD-controllers must be immediately reflected in the `ShootState` resource.
+The contract between Gardener and those controllers is: **Every controller must be capable of reconstructing its own environment based on both the state it has written before and on the real world's conditions/state.**
+
+```yaml
+---
+apiVersion: gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: shoot--core--aws-01
+shootRef:
+  name: aws-01
+  project: core
+state:
+  secrets:
+  - name: ca
+    data: ...
+  - name: kube-apiserver-cert
+    data: ...
+  resources:
+  - kind: DNS
+    name: record-1
+    state: <copied-state-of-dns-crd>
+  - kind: Infrastructure
+    name: networks
+    state: <copied-state-of-infrastructure-crd>
+  ...
+  <other fields required to keep track of>
+```
+
+We cannot assume that Gardener is always online to observe the most recent states the controllers have written to their resources.
+Consequently, the information stored here must not be used as "single point of truth", but the controllers must potentially check the real world's status to reconstruct themselves.
+However, this must anyway be part of their normal reconciliation logic and is a general best practice for Kubernetes controllers.
+
+#### Shoot health checks/conditions
+
+Some of the existing conditions already contain specific code which shall be simplified as well.
+All of the CRDs described above have a `.status.conditions` field to which the controllers may write relevant health information of their function area.
+Gardener will pick them up and copy them over to the Shoots `.status.conditions`.
```

Comment:
Can you explain your comment in more detail?

Reply:
Probably this is the same thing you've also already thought of - A shoot's status wont contain all conditions of e.g. its DNS 'child', it will only contain a single condition like DNSHealthy - yes or no.

Reply:
OK, and how should Gardener know which conditions to transport?

Reply:
ping @Adracus

Reply:
That has to be part of a contract, i.e. for each resource, there is a predefined set of condition types Gardener knows about and Gardener will use to transport to the Shoot.

Reply:
OK, agreed that it may make sense that a controller writes more conditions to its resource than Gardener wants to transport, and that we need a mechanism to decide which to transport. WDYT about having `propagate={true,false}` field (maybe better name) on each condition?

Reply:
I guess `propagate` or `aggregate` is okay, then the gardener would use these conditions and form one single condition on shoot side from those.

Reply:
OK, I included that part.